### ISSUE TITLE
Chutzpah takes a long time while scanning unnecessary files

### ISSUE THREAD

**Issue by arch-imp** [mmanela/chutzpah#395]
I have the following project structure:

```
. _ProjectRoot
+-- _gulp
+-- _js
|   +-- _tests
|   +-- jquery.js
+-- _node_modules
|   +-- _<big folder structure>
+-- chutzpah.json
```

Given the following chutzpah.json

```
{
    "Tests":
    [
        { "Path": "js/tests", "Includes": [ "*Tests.js" ] }
    ]
}
```

When I right-click on the project inside Visual Studio and select "Run JS Tests", the chutzpah.log indicates that Chutzpah is spending about 2 minutes just scanning the node_modules folder looking for settings files and trying to build a test context for every javascript file it finds.  Due to the filter in the settings file, it correctly discards them.

Given my settings file filters out any javascript files outside of js/tests, why is Chutzpah spending 2 minutes looking elsewhere?  The time to look in js/tests is only 10 seconds.  The total time for the test run is almost 4 minutes. Half the time needlessly scanning node_modules.

Ironically, when I run the same tests and folder structure under a TFS build with the Test Sources Spec set to `ProjectRoot\**\*Tests.js`, chutzpah never even looks outside js\tests for additional settings files or tries to build test contexts in node_modules.

Is there some configuration that I'm missing?  How can I tell Chutzpah not to scan the node_modules folder?

**Update:** I'm not very good at markup. I don't know how to draw a folder structure. Just to clarify, the items in my attempt with an underscore are to represent folders.  The underscore isn't actually part of the name.

**Comment by mmanela** [comment]
Thanks for reporting this. I know about this issue and want to address it. First, correct me if I am wrong. Chutzpah is only scanning the node_modules folder on first loading of the solution. After that if you click run tests it should not scan those again. If that is happening then this may not be what I think it is.

 There are two reason it does not yet work the way you expect.
1. Chutzpah VS adapter was written before I added the .json file to allow you to filter tests. I would like to change VS to REQUIRE a json file to tell it what to scan. However, that would be a big breaking change. I am still trying to figure out the right way to introduce this change without instantly breaking people. 

2. The way Chutzpah VS Adapter works now it can detect when a js file that was not a test file becomes one. With the change you mentioned this would not work the same way. There would need to be a new solution for this and maybe it is as simple as re-scanning the test files more often.

**Comment by arch-imp** [comment]
Apologies if I'm missing something, but I don't see how either of your two reasons should stand in the way of this working as I expect.

You don't need to require a json file to tell it what to scan.  If no file is present, it scans the whole project just like it does today.  It looks for any files that are tests. If a new test file appears, or an existing file is changed to include tests, Chutzpah can add those tests.

In my desired behavior, the json file tells chutzpah to only scan for test files where I tell it to.  If a test file exists outside of where I tell it, then I expect that test file to not be scanned and any tests in the file would be ignored.  So, for example, in my above configuration, if a test file appears somewhere under node_modules, I expect those tests to be ignored, therefore there is no reason to even begin scanning the node_modules folder.  If a json file exits with a valid "Tests" setting, then don't scan anything EXCEPT where configued in the json file, which in my case would be anything under js/tests.  If a json file does not exist, then assume there are no test filters and scan everything.

I don't believe this would break existing functionality.

As for the first part of your answer, chutzpah doesn't scan when the solution is opened. It doesn't begin to scan unless I open the Test Explorer window.  In my case, I don't open it, but instead choose "Run JS Tests" from the Chutzpah context menu (when right clicking the project).  In either case, it scans all folders attempting to build a test context for each file every time I run the tests.  It also does this scan on every test run.

**Comment by mmanela** [comment]
Thanks, I appreciate the time you are taking to work with me on this issue.

First, regarding how Chutzpah runs. After opening a solution __first__ time you run sets Chutzpah will run the TestContainerDiscoverer and add a file watch on all .js files in your solution. This is not ideal and what I want to change. But after it does that it will create "test containers" only for the files which match your chutzpah.json files. These test containers are the set of files which will be scanned for tests in the future.

From now on whenever you run tests Chutzpah will just hand the test adapter infrastructure those set of test containers. This will never contain the files not mapped from the Tests setting in the Chutzpah.json file, so this should be faster. 

So the optimization we are talking about should only improve the once time perf hit to scan all files as well as the memory pressure of added file watchers.


Second, You are right about the Chutzpah.json. I think it would be safe to pivot off the existence of it. It is a bit more complicated since you can have multiple chutzpah.json files and they can inherit but that can be handled.

**Comment by arch-imp** [comment]
I have run some scenarios to clarify what I'm experiencing and what you're saying should happen.  I ran 4 scenarios with a minimal project as described in my original post.
1- Open the solution then choose "Run JS Tests" from the chutzpah context menu
2- Open the solution then open Test Explorer and allow test discovery
3- Open the solution, open test explorer, then choose "Run JS Tests" from the chutzpah context menu
4- Open the solution, open test explorer, then choose "Run All" from the test explorer window.
What I have found leads me to believe that running tests from the "Run JS Tests" chutzpah context menu ignores any test discovery info collected from the test explorer

In scenarios 2 and 4, it appears that chutzpah will only look in folders that are actually part of the project.  In my above described sample project, the js and child folders are part of the project.  The node_modules folder is not.  Chutzpah will discover tests and run tests as specified in the chutzpah.json file.  It adds file watch to each javascript file found. Now, I question the usefulness of adding file watch to files that don't meet the the chutzpah.json file "Tests" setting, but I can live with that for now as it is a small time difference.  Chutzpah makes no reference to node_modules folder, so I think we're safe there.

In scenarios 1 and 3, it appears that chutzpah does not use any information from test discovery. What I see from the log file is chutzpah is scanning every folder from the project folder on down, including all folders not part of the project.  While scanning, it looks for chutzpah.json files and finds the one at the top level.  Then it again traverses every folder and examines every javascript file to attempt a test context.  While attempting to build a test context, chutzpah will reject the context if the file does not meet the "Tests" setting in chutzpah.json file.

What I would like changed is the scanning during scenarios 1 and 3.  I think that chutzpah should locate the top level chutzpah.json file, then ignore any file or folder that doesn't meet the "Tests" filter setting.  This includes looking for additional chutzpah.json files.(chutzpah.json files found along side tests would be acceptable.) This will eliminate any unnecessary scanning as I have already told Chutzpah.json where all my test files are located.  There is no value in scanning anywhere else.

I have not tired, but I would think running chutzpah from the commandline would behave as scenario 1 and 3.

Is there some value in scanning outside of the test filter when choosing the "Run JS Tests" chutzpah context menu?

I can share my project and the 4 log files but I wouldn't know where to post them for you to get.

**Comment by mmanela** [comment]
Ok, I know understand you better. I got confused since you were talking about two different Chutzpah extensions. You are correct that the "Context menu" extensions (as well as the command line client) will do an inefficient scanning if you tell it to run directly on a folder. However, it does exactly what you want if you run Chutzpah directly on the chutzpah.json file. I am facing a problem of accounting for what is technically possible versus what I think most people do. 

1. Chutzpah supports having a hierarchy of chutzpah.json files and stopping at the top most would not be correct.

2. A user from the context menu extension could choose a top level JS file AND a second JS file that is below a chutzpah.json file. I would need more complex logic to handle this.


As you pointed out though, in the end Chutzpah still will reject JS files that don't match the test settings.  But you are suggesting just doing that earlier. We would need to test to see how much performance gain this would add.

**Comment by arch-imp** [comment]
In our case, this is a huge performance gain. Scanning files that are subsequently rejected takes 2 min of a 4 min total test run.  That's 50% of the time wasted.  The more javascript libraries and 3rd party software I rely on, the worse it gets.

Although I can work around this by using Test Explorer or the commandline to run my tests, the Chutzpah context menu "Run JS Tests" now takes too much time to use with any frequency.  Which is sad, because that's my favorite way to run the tests.

**Comment by mmanela** [comment]
But it does work find if you do "Run JS Tests" on the chutzpah.json file?

**Comment by arch-imp** [comment]
"Run JS Tests" is not on the context menu when right-clicking on the chutzpah.json file.  Is it supposed to be? Is there another way to trigger the command from the chutzpah.json file?

"Run JS Tests" is only available on the project itself, folders, or .js files. I don't have any coffeescript or typescript files, so I can't comment on those.

**Comment by arch-imp** [comment]
Just to add more data to the reasons I'd like to see files ignored that are outside the "Tests" setting :)...
I've added "EnableTestFileBatching":"true" to my chutzpah.json file.
If I run all the tests from Test Explorer or the commandline (passing in the chutzpah.json file) the tests run.
If I use "Run JS Tests" from the tests folder, the tests run.
If I use "Run JS Tests" from the project node, the log indicates chutzpah is attempting to build a test context for all the files it found throughout the folder structure (including those in node_modules) and exits with the error: "Cannot build a batch context for Url or Html test files". (There are .html files deep in node_modules.)
If I use "Run JS Tests" from the js folder (one above the tests folder and includes the javascript file that is being tested) then Chutzpah exits with the error: "Failed to detect test framework for '<the javascript file that is being tested>'.

My conclusions are I need to be careful where I choose to execute "Run JS Tests" from.  I can't simply choose the project or any parent folders. Only the folder containing the tests works properly.

**Comment by mmanela** [comment]
Sure, I can look for ways to improve but as a fast fix would it help if you could just click on the .json file?

**Comment by mmanela** [comment]
Also, is the EnableTestFileBatching working well for you?

**Comment by arch-imp** [comment]
Like I said earlier, I can't run the tests by right click on the json file.  The menu option isn't available on the json file.  For now, I'm choosing to run Chutzpah via test explorer or the commandline or clicking on the tests folder.  I'm not looking for a fast fix at this time.  Just hoping for a future change.
EnableTestFileBatching works as long as I work around the tendency to scan folders outside my tests.

**Comment by mmanela** [comment]
So in short you are suggestion to assume that if you have **any** chutzpah.json with a tests section than ignore **all** other potential js files.

**Comment by arch-imp** [comment]
That's how I would expect it to work, yes.  Do you see potential problems making it work this way?

**Comment by olohmann** [comment]
Facing similar performance problems in a code base that has node_modules and bower_components folders. VS is getting quickly up to >1.7 GiB of RAM and 100% CPU. I would actually be happy with a simple flag that disables all scanning and lettings chutzpah just read the JSON config... I've automated the chutzpah.json file generation via gulp (using Karma at CI) to avoid manual maintenance of the file.

**Comment by Jonemode** [comment]
Hi team,
I am also experiencing huge performance hits (100% CPU usage for ~5minutes) due to my node_modules folder being scanned. I would like to opt for a solution that lets the user specify scanning behaviour via chutzpah.json, whether it be like what olohmann suggested or something similar. Other than this issue, the test runner performs great

**Comment by steve-tapley** [comment]
Hi,

+1 to the large performance hit at startup - its of the order of 5 minutes for us as well - the solutions proposed already would also work well for us.


We also don't see an option to "Run JS Tests" when right clicking on the chutzpah.json file.   In our case we have a TypeScript project (its still a .csproj file) that has been migrated from VS2013

Thanks,
Steve

**Comment by mmanela** [comment]
@steve-tapley I think you only have the adapter extension install. Chutzpah has two extensions, one is a context menu extension which add Chutzpah specific menus. Can you check that out and see if it hepls?

If you do have the adapter project installed and you are seeing a start up hit that usually means that it is scanning a lot of files. How large is your project? If you have a noce_modules folder is it added as a project file?

**Comment by mmanela** [comment]
@Jonemode Same quetions that I asked to @steve-tapley apply to you.

**Comment by Jonemode** [comment]
Thanks for the suggestion @mmanela , I have both extensions installed and can access the context menus fine in any csproj but I am using a njsproj which it doesn't look like there is support for yet.
Regarding node_modules folder being scanned - I did not have my node_modules folder included in my solution and it was as if it was still being looked at by the chutzpah extension. My solution itself is small, about 5MB without node_modules but about 70MB with node_modules included. Note that I had to convert my project to a csproj to test with chutzpah.

**Comment by mmanela** [comment]
That is interesting @Jonemode. As that is not the behavior I see. 
Using the Adapter plugin I only see the issue if I have the node_modules added. Does the issue happen to you the second you open the solution or only after you perform some action?

The bug for the njsproj sounds real. I can take a look at that.

**Comment by Jonemode** [comment]
@mmanela The UI locks up for me when I navigate to Test Explorer. If I don't touch the test explorer it seems to be fine. I am not 100% sure if it locks up because of my node_modules folder, but it seems like a safe bet considering the rest of my app is small - maybe there's more I can do to find out what kind of processing is being done while the UI locks up at the Test Explorer?

**Comment by steve-tapley** [comment]
@mmanela We have both extensions installed, so not sure what is going on there

We don't have a node_modules folder.   To give you an idea of the size of the project:
39 .csproj files (approx half contain just C# files, the others are 99% javascript)
6800 .js files (that number does include things like our bower_components folders & scripts folders both of which I try to remove in my various chutzpah.json files)
1000 .html files
1000 .cs

I tend to have a chutzpah.json file in the relevant js projects, but no solution wide one.   They pretty much all look like this:

> {
  "Framework": "jasmine",
  "TestFileTimeout": 10000,
  "CodeCoverageExecutionMode": "Never",
  "References": [
    {
      "Path": "scripts/jquery.js"
    },
    {
      "Path": "scripts/angular.js"
    },
    {
      "Path": "scripts/angular-route.js"
    },
    {
      "Path": "scripts/angular-mocks.js"
    },
    {
      "Path": "scripts/angular-moment.js"
    },
    {
      "Path": "dist/ADAPT.Common-globals.js"
    },
    {
      "Path": "dist/ADAPT.Common.js"
    }
  ],
  "Tests": [
    {
      "Path": "src",
      "Includes": [ "*.spec.js" ]
    }
  ]
}

**Comment by mmanela** [comment]
Interesting. Thanks for the details. I need to look into this more and see where the slowness is and how to fix.

In the meanwhile I just released 4.2.0 which allows you to run directly from a chutzpah.json file when using the context menu extension which hopefully will help for some of the scenarios above.

**Comment by roopkt** [comment]
Hi All,
I am experiencing the same issue.  
I have installed chutzpah extension for both Test Adapter for Test Explorer and Test Runner 
(4.3.3 version). In my case whenever I was running any single file test case I was seeing long delay because there was one more Drop Down called TESTS in visual studio output window which was trying to discover tests and was doing it for other files which are not tests files and keep failing showing bunch of errors making life very slow. 
In order to fix it I just disabled the Chutzpah Test Adapter for the Test Explorer. However, now I miss the facility to run test from VS TEST Explorer. If anyone has any neat solution please let me know.

**Comment by mmanela** [comment]
Does the context menu extension still work fine for you?

The adapter needs to change to be performant but I haven't had a chance to investigate what the work is. I feel it is a large change though.

**Comment by roopkt** [comment]
Hi Matthew,
Thanks for replying, yes my context menu extension is working fine.  Initially when I was running test using context menu option each time I was running it was discovering Tests for VS Test Explorer and was making stuff super slow. Now it is fast enough. I am running my tests using context menu option itself. 
My sole motivation to upgrade chutzpah to 4.3.3 was  to use Visual Studio Debugging option. Previously I was in Chutzpah 1 version. But now since I disabled the Tests Explorer adapter I don't see the JavaScript tests listed in VS Test Explorer which is sad. I loved to run / debug JavaScript code in visual studio directly.

**Comment by mmanela** [comment]
You can still run/debug from the context menu extension though.

**Comment by roopkt** [comment]
Hi Matthew,
Thanks for letting me know this. I just observed that there is one new entry in context menu called Run Chutzpah With and their when I select Debugger then it debugs the JavaScript file right in Visual Studio. Which is superb.  Thanks once again!

**Comment by vladistan** [comment]
+1 on this.  I am also experiencing the same problems.  With Visual Studio 2013 and Chutzpah 4.3.4   

I have both visual extensions installed,  context menu works fine,  but unit test explorer takes forever to scan for tests and completely freezing IDE for 5-10 minutes.   I do have quite a few dependencies installed in node_modules and bower_components.  That I'd like to be completely ignored by Chutzpah
